bee-protocol/src/workers/packets/heartbeat.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Heartbeat packet of the protocol.
+
+use crate::workers::packets::Packet;
+
+use std::ops::Range;
+
+const SOLID_MILESTONE_INDEX_SIZE: usize = 4;
+const PRUNED_INDEX_SIZE: usize = 4;
+const LATEST_MILESTONE_INDEX_SIZE: usize = 4;
+const CONNECTED_PEERS_SIZE: usize = 1;
+const SYNCED_PEERS_SIZE: usize = 1;
+const CONSTANT_SIZE: usize = SOLID_MILESTONE_INDEX_SIZE
+    + PRUNED_INDEX_SIZE
+    + LATEST_MILESTONE_INDEX_SIZE
+    + CONNECTED_PEERS_SIZE
+    + SYNCED_PEERS_SIZE;
+
+/// A packet that informs about the part of the tangle currently being fully stored by a node.
+/// This packet is sent when a node:
+/// - just got paired to another node;
+/// - did a snapshot and pruned away a part of the tangle;
+/// - solidified a new milestone;
+/// It also helps other nodes to know if they can ask it a specific message.
+#[derive(Clone)]
+pub(crate) struct HeartbeatPacket {
+    /// Index of the latest solid milestone.
+    pub(crate) solid_milestone_index: u32,
+    /// Pruned index.
+    pub(crate) pruned_index: u32,
+    /// Index of the latest milestone.
+    pub(crate) latest_milestone_index: u32,
+    /// Number of connected peers.
+    pub(crate) connected_peers: u8,
+    /// Number of synced peers.
+    pub(crate) synced_peers: u8,
+}
+
+impl HeartbeatPacket {
+    pub(crate) fn new(
+        solid_milestone_index: u32,
+        pruned_index: u32,
+        latest_milestone_index: u32,
+        connected_peers: u8,
+        synced_peers: u8,
+    ) -> Self {
+        Self {
+            solid_milestone_index,
+            pruned_index,
+            latest_milestone_index,
+            connected_peers,
+            synced_peers,
+        }
+    }
+}
+
+impl Packet for HeartbeatPacket {
+    const ID: u8 = 0x04;
+
+    fn size_range() -> Range<usize> {
+        (CONSTANT_SIZE)..(CONSTANT_SIZE + 1)
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Self {
+        let (bytes, next) = bytes.split_at(SOLID_MILESTONE_INDEX_SIZE);
+        let solid_milestone_index = u32::from_le_bytes(bytes.try_into().expect("Invalid buffer size"));
```

Comment:
I'm starting to like the `split_as_array_at` idea.

Reply:
same point about this being optimized. But the `expect` calls are sort of pointless in that case. Maybe we should omit them.